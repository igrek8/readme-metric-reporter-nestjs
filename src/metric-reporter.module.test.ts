import { INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { MetricReporter } from 'readme-metric-reporter';
import request from 'supertest';
import { MetricReporterModule } from './metric-reporter.module';

jest.mock('os', () => ({
  ...jest.requireActual('os'),
  type: jest.fn().mockReturnValue('darwin'),
  release: jest.fn().mockReturnValue('10.0.0'),
}));

jest.useFakeTimers();
jest.setSystemTime(new Date('2022-01-01'));

describe('MetricReporterModule', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        MetricReporterModule.register({
          reporter: new MetricReporter('apiKey'),
          collector: { collect: (_req, _res, metric) => ({ ...metric, _id: '1' }) },
          routes: ['*'],
          excludedRoutes: ['/health'],
        }),
      ],
    }).compile();
    app = moduleRef.createNestApplication();
    await app.listen(50000);
  });

  afterAll(async () => {
    await app.close();
  });

  it('tracks a request', async () => {
    const report = jest.spyOn(MetricReporter.prototype, 'report');
    report.mockImplementation(async () => {});
    await request(app.getHttpServer()).post('/404').send({ ping: 'text' });
    expect(report.mock.calls[0]?.[0]).toMatchSnapshot();
  });
});
