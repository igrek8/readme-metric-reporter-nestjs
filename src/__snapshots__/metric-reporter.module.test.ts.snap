// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricReporterModule tracks a request 1`] = `
{
  "_id": "1",
  "clientIPAddress": "::ffff:127.0.0.1",
  "group": {
    "id": "::ffff:127.0.0.1",
  },
  "request": {
    "log": {
      "creator": {
        "comment": "darwin/10.0.0",
        "name": "readme-metric-reporter-express",
        "version": "1.1.0",
      },
      "entries": [
        {
          "pageref": "http://127.0.0.1:50000/404",
          "request": {
            "headers": [
              {
                "name": "host",
                "value": "127.0.0.1:50000",
              },
              {
                "name": "accept-encoding",
                "value": "gzip, deflate",
              },
              {
                "name": "content-type",
                "value": "application/json",
              },
              {
                "name": "content-length",
                "value": "15",
              },
              {
                "name": "connection",
                "value": "close",
              },
            ],
            "httpVersion": "HTTP/1.1",
            "method": "POST",
            "postData": {
              "mimeType": "application/json",
              "text": "{"ping":"text"}",
            },
            "queryString": [],
            "url": "http://127.0.0.1:50000/404",
          },
          "response": {
            "content": {
              "mimeType": "application/json",
              "size": 281,
              "text": "{"statusCode":404,"message":"Cannot POST /404","error":"Not Found"}",
            },
            "headers": [
              {
                "name": "x-powered-by",
                "value": "Express",
              },
              {
                "name": "content-type",
                "value": "application/json; charset=utf-8",
              },
              {
                "name": "content-length",
                "value": "67",
              },
              {
                "name": "etag",
                "value": "W/"43-Rfemzan0q/U/9Rghim8W1DwIzRA"",
              },
            ],
            "status": 404,
            "statusText": "Not Found",
          },
          "startedDateTime": "2022-01-01T00:00:00.000Z",
          "time": 0,
        },
      ],
    },
  },
}
`;
